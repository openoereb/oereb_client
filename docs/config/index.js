var path = require('canonical-path');

var Package = require('dgeni').Package;

module.exports = new Package('oereb_client_doc', [
  require('dgeni-packages/ngdoc'),
  require('dgeni-packages/nunjucks')
])

  .processor(require('./processors/index-page'))
  .processor(require('./processors/version'))
  .processor(require('./processors/api-data'))
  .processor(require('./processors/guide-data'))

  .config(function (log, readFilesProcessor, writeFilesProcessor) {

    // Set the log level to 'info', switch to 'debug' when troubleshooting
    log.level = 'info';

    // Specify the base path used when resolving relative paths to source and output files
    readFilesProcessor.basePath = path.resolve(__dirname, '../..');

    // Specify our source files that we want to extract
    readFilesProcessor.sourceFiles = [
      { // All of our application files
        include: 'oereb_client/static/js/**/*.js',
        basePath: 'oereb_client/static/js'
      },
      { // Our static Markdown documents
        include: 'docs/content/**/*.md',
        basePath: 'docs/content',
        fileReader: 'ngdocFileReader'
      }
    ];

    // Use the writeFilesProcessor to specify the output folder for the extracted files
    writeFilesProcessor.outputFolder = 'docs/build';

  })

  .config(function (parseTagsProcessor, getInjectables) {
    parseTagsProcessor.tagDefinitions =
      parseTagsProcessor.tagDefinitions.concat(getInjectables(require('./tag-defs')));
  })

  .config(function (templateFinder) {

    // Specify where the templates are located
    templateFinder.templateFolders.unshift(path.resolve(__dirname, 'templates'));

  })

  .config(function (computePathsProcessor, computeIdsProcessor) {

    computePathsProcessor.pathTemplates.push({
      docTypes: ['module'],
      pathTemplate: '${area}/${name}',
      outputPathTemplate: 'partials/${area}/${name}.html'
    });

    computePathsProcessor.pathTemplates.push({
      docTypes: ['componentGroup'],
      pathTemplate: '${area}/${moduleName}/${groupType}',
      outputPathTemplate: 'partials/${area}/${moduleName}/${groupType}.html'
    });

    // create new compute for 'content' type doc
    // indexPage is something new we will be defining later
    computeIdsProcessor.idTemplates.push({
      docTypes: ['content', 'indexPage'],
      getId: function (doc) {
        return doc.fileInfo.baseName;
      },
      getAliases: function (doc) {
        return [doc.id];
      }
    });

    // Build custom paths and set the outputPaths for "content" pages
    computePathsProcessor.pathTemplates.push({
      docTypes: ['content'],
      getPath: function (doc) {
        var docPath = path.dirname(doc.fileInfo.relativePath);
        if (doc.fileInfo.baseName !== 'index') {
          docPath = path.join(docPath, doc.fileInfo.baseName);
        }
        return docPath;
      },
      outputPathTemplate: 'partials/${path}.html'
    });

  })

  .config(function (extractAccessTransform) {
    extractAccessTransform.allowedDocTypes.add('js');
  });
