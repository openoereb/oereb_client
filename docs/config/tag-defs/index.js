module.exports = [
  require('./export'),
  require('./ngInject')
];