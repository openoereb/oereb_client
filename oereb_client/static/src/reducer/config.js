import {createSlice} from "@reduxjs/toolkit";

export const configSlice = createSlice({
  name: 'config',
  initialState: {
    config: {}
  },
  reducers: {
    update: (state, action) => {
      state.config = action.payload;
    }
  }
});

export const {update} = configSlice.actions;

export default configSlice.reducer;
