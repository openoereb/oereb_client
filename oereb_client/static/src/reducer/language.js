import {createSlice} from "@reduxjs/toolkit";

import i18n from "../i18n";

export const languageSlice = createSlice({
  name: 'language',
  initialState: {
    current: null,
    default: null,
    available: []
  },
  reducers: {
    initLanguages: (state, action) => {
      state.default = action.payload.default;
      state.available = action.payload.available;
      const query = new URLSearchParams(window.location.search);
      if (query.has('lang') && state.available.indexOf(query.get('lang')) > -1) {
        state.current = query.get('lang');
      }
      else {
        state.current = state.default;
      }
      i18n.changeLanguage(state.current);
      query.set('lang', state.current);
      window.history.pushState(null, null, '?' + query.toString());
    },
    setLanguage: (state, action) => {
      if (state.available.indexOf(action.payload) > -1) {
        state.current = action.payload;
        i18n.changeLanguage(state.current);
        const query = new URLSearchParams(window.location.search);
        query.set('lang', state.current);
        window.history.pushState(null, null, '?' + query.toString());
      }
    }
  }
});

export const {initLanguages, setLanguage} = languageSlice.actions;

export default languageSlice.reducer;
