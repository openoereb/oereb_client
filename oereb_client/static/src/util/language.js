import {isArray} from "lodash";

export const getLocalizedText = function (multilingualText, language, defaultLanguage) {
  if (isArray(multilingualText) && multilingualText.length > 0) {
    let defaultValue = null;
    for (let i = 0; i < multilingualText.length; i++) {
      if (multilingualText[i]["Language"] === language) {
        return multilingualText[i]["Text"];
      }
      else if (multilingualText[i]["Language"] === defaultLanguage) {
        defaultValue = multilingualText[i]["Text"];
      }
    }
    if (defaultValue) {
      return defaultValue;
    }
    return multilingualText[0]["Text"];
  }
  return null;
};

export const getLocalizedUrl = function (multilingualUrl, language, defaultLanguage) {
  if (isArray(multilingualUrl) && multilingualUrl.length > 0) {
    let defaultValue = null;
    for (let i = 0; i < multilingualUrl.length; i++) {
      if (multilingualUrl[i]["Language"] === language) {
        return multilingualUrl[i]["URL"];
      }
      else if (multilingualUrl[i]["Language"] === defaultLanguage) {
        defaultValue = multilingualUrl[i]["URL"];
      }
    }
    if (defaultValue) {
      return defaultValue;
    }
    return multilingualUrl[0]["URL"];
  }
  return null;
};
