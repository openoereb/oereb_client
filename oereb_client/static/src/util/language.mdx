import {Meta} from '@storybook/addon-docs/blocks';

<Meta title="API Reference/Util/Language" />

# getLocalizedText

Accepts a multilingual array containing localized texts for certain languages,
the requested language and the default language. The localized text of the
requested language is returned as result. If the requested language is not
available in the array, the default language is used as fallback.

```js
getLocalizedText(multilingualText, language, defaultLanguage);
```

### Parameters

| Param | Type | Description |
| --- | --- | --- |
| multilingualText | <code>Array</code> | The array containing the localized texts. |
| language | <code>String</code> | The requested language. |
| defaultLanguage | <code>String</code> | The default language used as fallback. |

### Returns

<code>String</code> The localized text.

# getLocalizedUrl

Accepts a multilingual array containing localized URLs for certain languages,
the requested language and the default language. The localized URL of the
requested language is returned as result. If the requested language is not
available in the array, the default language is used as fallback.

```js
getLocalizedUrl(multilingualUrl, language, defaultLanguage);
```

### Parameters

| Param | Type | Description |
| --- | --- | --- |
| multilingualUrl | <code>Array</code> | The array containing the localized URLs. |
| language | <code>String</code> | The requested language. |
| defaultLanguage | <code>String</code> | The default language used as fallback. |

### Returns

<code>String</code> The localized URL.
