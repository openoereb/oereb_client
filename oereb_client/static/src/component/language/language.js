import PropTypes from 'prop-types';
import React from "react";
import {withNamespaces} from 'react-i18next';
import {useDispatch, useSelector} from "react-redux";

import {setLanguage} from "../../reducer/language";

const OerebLanguage = function({t}) {
  const dispatch = useDispatch();
  const language = useSelector((state) => state.language);
  const currentLanguage = language.current;
  const availableLanguages = language.available;

  const selectLanguage = function(lang) {
    dispatch(setLanguage(lang));
  };

  const languages = availableLanguages.map((language, key) =>
    <li key={key}>
      <button className="dropdown-item" onClick={selectLanguage.bind(this, language)}>
        {language.toUpperCase()}
      </button>
    </li>
  );

  return (
    <div>
      <button className="btn btn-outline-secondary dropdown-toggle"
        type="button"
        data-bs-toggle="dropdown"
        aria-expanded="false"
        title={t('menu.language.title')}>
        {currentLanguage.toUpperCase()}
      </button>
      <ul className="dropdown-menu">
        {languages}
      </ul>
    </div>
  );
};

OerebLanguage.propTypes = {
  t: PropTypes.func.isRequired
};

export default withNamespaces()(OerebLanguage);
