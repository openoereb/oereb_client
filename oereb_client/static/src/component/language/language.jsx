import React from "react";
import {useTranslation} from "react-i18next";
import {useDispatch, useSelector} from "react-redux";

import {hideExtract, loadExtract, showError, showExtract} from "../../reducer/extract";
import {updateHistory} from "../../reducer/history";
import {setLanguage} from "../../reducer/language";
import {queryExtractById} from "../../request/extract";
import {TooManyRequestsError} from "../../util/error";
import {showWarning} from "../../reducer/message";

/**
 * This component provides the language selector.
 */
const OerebLanguage = function() {
  const {t} = useTranslation();
  const dispatch = useDispatch();
  const config = useSelector((state) => state.config).config;
  const serviceUrl = config.service_url;
  const language = useSelector((state) => state.language);
  const currentLanguage = language.current;
  const availableLanguages = language.available;
  const extract = useSelector((state) => state.extract);

  const selectLanguage = function(lang) {
    if (lang !== currentLanguage) {
      dispatch(setLanguage(lang));
      if (extract.visible) {
        dispatch(loadExtract({
          egrid: extract.egrid,
          identdn: extract.identdn,
          number: extract.number,
          zoom: false
        }));
        queryExtractById(
          serviceUrl,
          extract.egrid,
          extract.identdn,
          extract.number,
          config.extract_json_timeout,
          lang
        )
          .then((data) => {
            dispatch(showExtract(data));
            dispatch(updateHistory(data));
          })
          .catch((error) => {
            if (error instanceof TooManyRequestsError) {
              dispatch(hideExtract());
              dispatch(showWarning(t('extract.error.too_many_requests'), true));
            }
            else {
              dispatch(showError());
            }
          });
      }
    }
  };

  const languages = availableLanguages.map((language, key) =>
    <li key={key}>
      <button className="dropdown-item" onClick={selectLanguage.bind(this, language)}>
        {language.toUpperCase()}
      </button>
    </li>
  );

  return (
    <div className="oereb-client-language-selector">
      <button className="btn btn-outline-secondary dropdown-toggle"
        type="button"
        data-bs-toggle="dropdown"
        aria-expanded="false"
        title={t("menu.language.title")}>
        {currentLanguage.toUpperCase()}
      </button>
      <ul className="dropdown-menu">
        {languages}
      </ul>
    </div>
  );
};

export default OerebLanguage;
