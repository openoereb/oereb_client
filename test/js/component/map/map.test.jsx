import {act, render} from '@testing-library/react';
import {register} from "ol/proj/proj4";
import TileWMS from 'ol/source/TileWMS';
import WMTS from 'ol/source/WMTS';
import proj4 from "proj4";
import React from "react";
import {Provider} from "react-redux";

import OerebMap, {
  getBaseLayerSourceWms,
  getBaseLayerSourceWmts
} from "../../../../oereb_client/static/src/component/map/map";
import {update} from "../../../../oereb_client/static/src/reducer/config";
import {initLanguages} from "../../../../oereb_client/static/src/reducer/language";
import MainStore from "../../../../oereb_client/static/src/store/main";
import capabilities from "./assets/capabilities";

const config = {
  view: {
    center: [2615000.0, 1255000.0],
    zoom: 2,
    resolutions: [250, 100, 50, 20, 10],
    projection: 'EPSG:2056'
  },
  base_layer: {
    type: 'wms',
    url: 'http://example.com/wms',
    params: {
      LAYERS: 'baselayer'
    }
  },
  availability: {
    url: 'http://example.com/wms',
    params: {
      LAYERS: 'availability'
    }
  },
  mask_surrounding: {
    url: 'http://example.com/wms',
    params: {
      LAYERS: 'outside_bl_area',
      FORMAT: 'image/png'
    },
    opacity: 0.8
  },
  use_tile_wms: false
};

beforeEach(() => {
  fetch.resetMocks();
  proj4.defs(
    'EPSG:2056',
    '+proj=somerc +lat_0=46.95240555555556 +lon_0=7.439583333333333 +k_0=1 +x_0=2600000 +y_0=1200000 ' +
    '+ellps=bessel +towgs84=674.374,15.056,405.346,0,0,0,0 +units=m +no_defs'
  );
  register(proj4);
});

describe('map component', () => {

  let component;

  beforeEach(() => {
    act(() => {
      MainStore.dispatch(initLanguages({
        default: 'de',
        available: ['de']
      }));
    });
    act(() => {
      MainStore.dispatch(update(config));
    });
    component = render(
      <Provider store={MainStore}>
        <OerebMap />
      </Provider>
    );
  });

  it('should render map', () => {
    expect(component.asFragment()).toMatchSnapshot();
  });

});

describe('getBaseLayerSourceWms', () => {

  it('should return the WMS source', async () => {
    const source = await getBaseLayerSourceWms({
      url: 'http://example.com/wms',
      params: {
        LAYERS: 'grundkarte_farbig',
        FORMAT: 'image/png'
      },
      projection: 'EPSG:2056',
      attributions: 'baselayer wms'
    });
    expect(source).toBeInstanceOf(TileWMS);
    expect(source.getParams()['LAYERS']).toEqual('grundkarte_farbig');
    expect(source.getParams()['FORMAT']).toEqual('image/png');
    expect(source.getProjection().getCode()).toEqual('EPSG:2056');
    expect(source.getAttributions()()[0]).toEqual('baselayer wms');
  });

});

describe('getBaseLayerSourceWmts', () => {

  it('should return the WMTS source', async () => {
    fetch.mockResponseOnce(capabilities);
    const source = await getBaseLayerSourceWmts({
      url: 'http://example.com/wmts',
      layer: 'grundkarte_farbig',
      matrixSet: 'swissgrid',
      projection: 'EPSG:2056',
      style: 'default',
      format: 'image/png',
      attributions: 'baselayer wmts'
    });
    expect(source).toBeInstanceOf(WMTS);
    expect(source.getLayer()).toEqual('grundkarte_farbig');
    expect(source.getMatrixSet()).toEqual('swissgrid');
    expect(source.getProjection().getCode()).toEqual('EPSG:2056');
    expect(source.getStyle()).toEqual('default');
    expect(source.getFormat()).toEqual('image/png');
    expect(source.getAttributions()()[0]).toEqual('baselayer wmts');
    expect(fetch.mock.calls).toHaveLength(1);
  });

});
