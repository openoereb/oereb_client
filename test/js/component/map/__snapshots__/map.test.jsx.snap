// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`map component > should render map 1`] = `
<DocumentFragment>
  <div
    class="oereb-client-map-wrapper"
  >
    <div
      class="oereb-client-map"
    >
      <div
        class="ol-viewport ol-touch"
        style="position: relative; overflow: hidden; width: 100%; height: 100%;"
      >
        <div
          class="ol-unselectable ol-layers"
          style="position: absolute; width: 100%; height: 100%; z-index: 0;"
        />
        <div
          class="ol-overlaycontainer"
          style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;"
        />
        <div
          class="ol-overlaycontainer-stopevent"
          style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;"
        >
          <div
            class="ol-overlay-container ol-selectable"
            style="position: absolute; pointer-events: auto; display: none;"
          >
            <div
              class="oereb-client-overlay"
            />
          </div>
          <div
            class="ol-zoom ol-unselectable ol-control"
            style="pointer-events: auto;"
          >
            <button
              class="ol-zoom-in"
              title="Zoom in"
              type="button"
            >
              +
            </button>
            <button
              class="ol-zoom-out"
              title="Zoom out"
              type="button"
            >
              â€“
            </button>
          </div>
          <div
            class="ol-attribution ol-unselectable ol-control ol-collapsed"
            style="pointer-events: auto;"
          >
            <button
              aria-expanded="false"
              title="Attributions"
              type="button"
            >
              <span
                class="ol-attribution-expand"
              >
                i
              </span>
            </button>
            <ul />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
